feat: captcha auto-restart

Airwave1981 opened the pull request:
## ℹ️ Description
Adds an **automatic restart loop after a captcha is encountered**.  
Instead of waiting for manual confirmation, the bot now

1. aborts the current run as soon as a captcha appears,
2. sleeps for the duration defined in `captcha.restart_delay_h` (default 6 h),
3. starts a fresh run automatically.

This allows continuous publishing of up to 10 ads, then a cooldown, then another 10, … without user interaction.

*Related issue:* #<issue-number>

## 📋 Changes Summary
- **`src/kleinanzeigen_bot/__main__.py`**  
  * new while-loop wrapper that catches `CaptchaEncountered`, sleeps, and restarts  
  * YAML helper to read `captcha.restart_delay_h` from config
- **`src/kleinanzeigen_bot/__init__.py`**  
  * (optional) added `shutdown()` coroutine for graceful browser / WebSocket cleanup
- **`src/kleinanzeigen_bot/utils/exceptions.py`**  
  * exports `CaptchaEncountered` for external use
- **`src/kleinanzeigen_bot/resources/translations.de.yaml` | Added German translation for new log message |
- **`README.md`**  
  * documented the new `restart_delay_h` option and the auto-restart behaviour

### ⚙️ Type of Change
- [ ] 🐞 Bug fix
- [x] ✨ New feature
- [ ] 💥 Breaking change

## ✅ Checklist
- [x] I have reviewed my changes to ensure they meet the project's standards.
- [x] I have tested my changes and ensured that all tests pass (`pdm run test`).
- [x] I have verified that linting passes (`pdm run lint`).
- [x] I have run security scans and addressed any identified issues (`pdm run audit`).
- [x] I have updated documentation where necessary.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

sebthom commented on tests/unit/test_i18n.py:
```diff
@@ -25,7 +27,7 @@ def test_detect_locale(monkeypatch: MonkeyPatch, lang: str | None, expected: i18
 
     # Call the function and compare the result to the expected output.
-    result = i18n._detect_locale()  # pylint: disable=protected-access
+    result = i18n._detect_locale()
```
can you please revert that change. I'd prefer to disable pylint rules locally and not globally for a fi.e

Heavenfighter commented on src/kleinanzeigen_bot/__init__.py:
```diff
@@ -15,6 +15,7 @@
 import certifi, colorama, nodriver
 from ruamel.yaml import YAML
 from wcmatch import glob
+from kleinanzeigen_bot.utils.exceptions import CaptchaEncountered
```
Can you harmonize this import along to the other utils imports? 
`from .utils.exceptions import captchaEncountered`

Airwave1981 replied:
I don't know the hell. Working locally no problem, if I commit - here I have build errors. Guess thats the reason why also the main is not relative currently.

Heavenfighter commented on src/kleinanzeigen_bot/__main__.py:
```diff
@@ -4,6 +4,47 @@
-kleinanzeigen_bot.main(sys.argv)
+    except CaptchaEncountered:
+        delay_h = _read_restart_delay()
+        print(f"[INFO] Captcha detected. Sleeping {delay_h} h before restart…")
```
Maybe add an translation for this text

Heavenfighter reviewed (commented):
Code looks god so far as i can see. I couldnt test  because i have little time and couldnt wait an hour.

Heavenfighter commented:
Why are there commits every minute?

Airwave1981 commented:
> Why are there commits every minute?

Tried to get the relative path fixed, as you wanted, but it is not building online, just locally. Guess thats the reason why also main iss not relative

Heavenfighter reviewed (dismissed):
Looks good to me

Airwave1981 commented:
> > > Why are there commits every minute?
> > 
> > 
> > Tried to get the relative path fixed, as you wanted, but it is not building online, just locally. Guess thats the reason why also main iss not relative
> 
> Oh ok, so pdm run test works locally?

Yes.

sebthom commented on src/kleinanzeigen_bot/__main__.py:
```diff
@@ -4,6 +4,49 @@
+while True:
+    try:
+        kleinanzeigen_bot.main(sys.argv)          # runs & returns when finished
+        sys.exit(0)                               # prevents closing issues
```
doesn't this terminate the process and thus no looping occurs?

sebthom commented:
@Airwave1981 @Heavenfighter I simplified the implementation a bit and avoided reparsing the yaml file. can you please check if the code still works for you as expected. I also changed captcha.restart_delay_h to captcha.restart_delay and you can no specify a duration like "4h 20m 4s"

Airwave1981 commented:
> @Airwave1981 @Heavenfighter I simplified the implementation a bit and avoided reparsing the yaml file. can you please check if the code still works for you as expected. I also changed captcha.restart_delay_h to captcha.restart_delay and you can no specify a duration like "4h 20m 4s"

Thanks a lot. Everything still works fine.
Just added a little change as I recognized, that the logger was instantiated multiple times.
